test/e2e/helpers_test.go

```diff
@@ -0,0 +1,285 @@
+func getEtcdMemberListPB(t *testing.T, c *envconf.Config, podName string) *etcdserverpb.MemberListResponse {
+	t.Helper()
+	stdout, stderr, err := execInPod(t, c, podName, namespace, []string{"etcdctl", "member", "list", "-w", "json"})
+	if err != nil {
+		t.Fatalf("Failed to get etcd member list: %v, stderr: %s", err, stderr)
+	}
+	var memberList etcdserverpb.MemberListResponse
+	if err := json.Unmarshal([]byte(stdout), &memberList); err != nil {
+		t.Fatalf("Failed to parse etcd member list JSON: %v", err)
+	}
+	return &memberList
+}
+
+// waitForNoLearners waits until the member list has the expected number of members
+// and all members are voting (i.e., no learners remain).
+func waitForNoLearners(t *testing.T, c *envconf.Config, podName string, expectedMembers int) {
+	t.Helper()
+	err := wait.For(func(ctx context.Context) (bool, error) {
+		ml := getEtcdMemberListPB(t, c, podName)
+		if len(ml.Members) != expectedMembers {
+			return false, nil
+		}
+		for _, m := range ml.Members {
+			if m.IsLearner {
+				return false, nil
+			}
+		}
+		return true, nil
+	}, wait.WithTimeout(3*time.Minute), wait.WithInterval(5*time.Second))
+	if err != nil {
+		t.Fatalf("Timeout waiting for %d voting members with no learners: %v", expectedMembers, err)
+	}
+}
+
+// verifyPodUsesPVC checks that a pod is using PVC for persistent storage
+func verifyPodUsesPVC(t *testing.T, c *envconf.Config, podName string, expectedPVCPrefix string) {
+	t.Helper()
+	var pod corev1.Pod
+	if err := c.Client().Resources().Get(t.Context(), podName, namespace, &pod); err != nil {
+		t.Fatalf("Failed to get pod %s: %v", podName, err)
+	}
+
+	// Check for PVC volumes
+	hasPVC := false
+	for _, volume := range pod.Spec.Volumes {
+		if volume.PersistentVolumeClaim != nil {
+			if strings.HasPrefix(volume.PersistentVolumeClaim.ClaimName, expectedPVCPrefix) {
+				hasPVC = true
+				break
+			}
+		}
+	}
+
+	if !hasPVC {
+		t.Errorf("Pod %s does not use expected PVC with prefix %s", podName, expectedPVCPrefix)
+	}
+}
+
+// getClusterEndpointHashKVs executes `etcdctl endpoint hashkv --cluster -w json` inside the given pod
+// and returns the parsed HashKV responses from all known endpoints using etcd's native types.
+func getClusterEndpointHashKVs(t *testing.T, c *envconf.Config, podName string) []etcdserverpb.HashKVResponse {
+	t.Helper()
+	cmd := []string{"etcdctl", "endpoint", "hashkv", "--cluster", "-w", "json"}
+	stdout, stderr, err := execInPod(t, c, podName, namespace, cmd)
+	if err != nil {
+		t.Fatalf("Failed to get cluster endpoint hashkv from %s: %v, stderr: %s", podName, err, stderr)
+	}
+
+	// Expected JSON: array of objects like {"Endpoint":"...","HashKV":{...}}
+	var entries []struct {
+		Endpoint string                      `json:"Endpoint"`
+		HashKV   etcdserverpb.HashKVResponse `json:"HashKV"`
+	}
+	if err := json.Unmarshal([]byte(stdout), &entries); err != nil {
+		t.Fatalf("Failed to parse endpoint hashkv JSON: %v. Raw: %s", err, truncate(stdout, 512))
```

Comment:
print the whole response? It won't too long.

Reply:
done